### PR TITLE
Add WebNN baseline implementation for first-wave ops

### PR DESCRIPTION
**Pull request by huningxin**
This PR implements 41 WebNN ops for [first-wave models](https://github.com/webmachinelearning/webnn/blob/main/op_compatibility/first_wave_models.md) in JavaScript double-precision Number and calculations without depending on 3rd party libs. It could be used to generate the baseline results for WebNN conformance tests.

Fix https://github.com/webmachinelearning/webnn/issues/245

@anssiko @dontcallmedom @wchao1115 @pyu10055 , PTAL.

/cc @BruceDai

### CODE REVIEW

**Review by anssiko — APPROVED** [review]
Thanks @huningxin and @BruceDai for this significant work. This proposal passes my review from the procedural perspective:

- Per our discussion on [WebML WG Teleconference – 2 Dec 2021](https://www.w3.org/2021/12/02-webmachinelearning-minutes.html#t03), webnn-baseline helps the WebML WG demonstrate Web Neural Network API implementation experience, a key requirement for advancing to a Candidate Recommendation.

- Considering the above and per [WebML CG Charter: Test Suites and Other Software](https://webmachinelearning.github.io/charter/#test-suites), webnn-baseline is now [proposed](https://www.w3.org/2022/01/27-webmachinelearning-minutes.html#t16) as a new CG deliverable. (The CG and WG will continue to work closely together per [Coordination](https://webmachinelearning.github.io/charter/#coordination).)

**Comment by dontcallmedom** [comment]
Great work! In order to help with the goal of this codebase (making it easy to review it), I've [created separately a couple of additional commits](https://github.com/huningxin/webnn-baseline-1/pull/1) that I think help in that direction.

They take the following steps:
* move any code that isn't directly a WebNN operation (but rather support code for operations) into the `src/lib` directory
* tersify the definitions of operations by
  * moving as much of the parameters normalization into the function signature using JavaScript deconstruction patterns (and other syntactical shortcuts)
  * moving all the validation of the parameters into a separate module, to have the main module focus only on the actual calculation

An alternative to the latter subbullet might be to get rid of parameters validation altogether - since the code isn't actually meant to be used in any production environment, it's not clear that the code adds a lot of value. The refactoring would make that a fairly easy additional change.

**Comment by huningxin** [comment]
Thanks much @dontcallmedom !

> * move any code that isn't directly a WebNN operation (but rather support code for operations) into the `src/lib` directory

+1. In the implementation, some big ops are implemented by composing small ops. These small ops are also WebNN ops, like `reshape`, so please keep them in the src folder.

> * moving as much of the parameters normalization into the function signature using JavaScript deconstruction patterns (and other syntactical shortcuts)

+1. I like that.

> * moving all the validation of the parameters into a separate module, to have the main module focus only on the actual calculation

+1

> An alternative to the latter subbullet might be to get rid of parameters validation altogether - since the code isn't actually meant to be used in any production environment, it's not clear that the code adds a lot of value.

It's a fair point. 

However, I suppose the validation code would help test developer to verify the test cases. @BruceDai 

Another perspective is it probably could help the spec editors to write the validation steps of a method, as https://github.com/webmachinelearning/webnn/issues/210. @anssiko @wchao1115

**Comment by dontcallmedom** [comment]
> > An alternative to the latter subbullet might be to get rid of parameters validation altogether - since the code isn't actually meant to be used in any production environment, it's not clear that the code adds a lot of value.
> 
> However, I suppose the validation code would help test developer to verify the test cases. @BruceDai

I can certainly relate to that - they did help mind find and fix bugs I introduced in the process of refactoring the code :)
 
> Another perspective is it probably could help the spec editors to write the validation steps of a method, as [webmachinelearning/webnn#210](https://github.com/webmachinelearning/webnn/issues/210). @anssiko @wchao1115

indeed

**Comment by anssiko** [comment]
If the editors feel the validation code helps them translate checks into corresponding spec validation steps, it sounds like a good reason to keep that part to facilitate this work. A separate validation module is a good idea too from a code organization perspective.

**Comment by huningxin** [comment]
https://github.com/huningxin/webnn-baseline-1/pull/1 merged. Thanks @dontcallmedom 's contribution that improves the reviewability much.

File: src/conv2d.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+import {transpose} from './transpose.js';
+
+/**
+ * Compute a 2-D convolution given 4-D input and filter tensors.
+ * @param {Tensor} input
+ * @param {Tensor} filter
+ * @param {MLConv2dOptions} options
+ * @return {Tensor}
+ */
+export function conv2d(input, filter, {padding = [0, 0, 0, 0],
```
Comment (huningxin) [inline src/conv2d.js:15]: @wchao1115, this function implements the steps of conv2d, just for your convivence, thanks!

**Comment by anssiko** [comment]
@huningxin [as per our discussion](https://www.w3.org/2022/05/05-webmachinelearning-minutes.html#t11), please feel free to merge this PR now to unblock https://github.com/webmachinelearning/webnn/issues/265

Thanks to all reviewers and special thanks to @huningxin and @BruceDai for preparing this initial contribution that is an important deliverable to demonstrate implementation experience and satisfy interoperability requirements.

**Pull request merged**